### PR TITLE
Raise NotImplementedError when SNI not present.

### LINKED ISSUE #458: SSL_set_tlsext_host_name may not be present
https://github.com/pyca/pyopenssl/blob/master/src/OpenSSL/SSL.py#L1239 assumes the binding is available, but < 1.0.0 that is not true. We're dropping support for 0.9.8 in cryptography shortly, but we should probably fix this for pyOpenSSL 16.1 anyway.

This manifests as an AttributeError when using requests + pyOpenSSL in CentOS 5. /cc @Lukasa

### PR DESCRIPTION
**Pull request by Lukasa**
Resolves #458.

@hynek, as discussed, this currently adds no test coverage for this new code. Do you want to see that added?

### CODE REVIEW

**Comment by hynek** [comment]
Does this actually fix urllib3 or does it just break in a…more obvious way?

**Comment by Lukasa** [comment]
It breaks it in a more obvious way. =)

I think dealing with this problem is not PyOpenSSL's job: there's really nothing it can do but fail appropriately. But I think we should object properly: `AttributeError` is a weird thing to raise here.

**Comment by hynek** [comment]
OK…so…I really don’t care enough.  Sorry. :)

If urllib3 has to catch an error anyways, they can just as well catch this AttributeError and introspect.  And while it it’s perfect, the error kind of makes sense: a method is missing. :)

cryptography is dropping 0.9.8 with its next release, it doesn’t make any sense to add all this code to pyOpenSSL, just to rip it out again later this year.  Agreed @reaperhulk ?

***

While looking at it I realized tho is that our `_requires_alpn` adds unnecessary runtime overhead.

IMHO it should be more like:

```python
def _requires_alpn(func):
   if not haz_alpn:
        @wraps(func)
        def explode(*args, **kw):
            raise NotImplementedErrror(yaddayadda)
        return explode

    return func
```

right?

**Comment by reaperhulk** [comment]
We are dropping 0.9.8 in cryptography 1.4, but with a flag to turn it back on. Our assumption is that this will be okay and that in 1.5 we'll completely remove all support.

**Comment by hynek** [comment]
blergh ugh.

so Cory, please make the decorator like I outlined above and add some monkey patched tests after you’ve tested it actually works. *sigh*

you also have my permission to fix _requires_alpn while doing so because it’s totally related. ;)

**Comment by hynek** [comment]
Maybe we should even create a generic solution since such problems will haunt us forever…

```python
_requires_sni = _make_requires(_lib.Cryptography_HAS_TLSEXT_HOSTNAME, "SNI not available")
```

maybe?

(n.b. I wouldn’t make assumptions *why* it’s not available because it might just as well be another asinine packaging problem)

**Comment by Lukasa** [comment]
Yeah, I think that's a reasonable approach.

**Comment by Lukasa** [comment]
@hynek Considering that this new code detects the features at import time, what's your suggested method of testing the SNI stuff (other than being hopeful that the fact that it uses identical code to the NPN/ALPN side means it works)?

**Comment by hynek** [comment]
I’d say test it explicitly.  i.e.

```
class TestRequires(object):
    def test_met():
        pass
    def test_not_met():
        pass
```

Then just apply the function within the test methods.

**Comment by Lukasa** [comment]
What are you talking about codecov? If I hit all of the diff, how can I possibly have reduced coverage?

**Comment by hynek** [comment]
codecov is really unreliable lately :(((

File: tests/test_ssl.py [hunk 1]
```diff
@@ -3855,5 +3856,43 @@ def test_integers(self):
             self.assertTrue(isinstance(const, int))
 
 
+class TestRequires(TestCase):
+    """
+    Tests for the decorator factory used to conditionally raise
+    NotImplementedError when older OpenSSL's are used.
```
Comment (hynek) [inline tests/test_ssl.py]: dat apostrophe :)

File: tests/test_ssl.py [hunk 2]
```diff
@@ -3855,5 +3856,43 @@ def test_integers(self):
             self.assertTrue(isinstance(const, int))
 
 
+class TestRequires(TestCase):
```
Comment (hynek) [inline tests/test_ssl.py]: you wrote it as Java tests just to upset me, didn’t you? :)

**Comment by Lukasa** [comment]
Regardless, @hynek, this should be R4R now.

**Comment by Lukasa** [comment]
Oh wait, you did review it. NVM, more patches!

File: tests/test_ssl.py [hunk 3]
```diff
@@ -3874,8 +3874,8 @@ def inner():
             results.append(True)
             return True
 
-        self.assertTrue(inner())
-        self.assertEqual(results, [True])
+        assert inner()
+        assert results == [True]
```
Comment (hynek) [inline tests/test_ssl.py]: we do expected first…makes them nicely aligned

File: tests/test_ssl.py [hunk 4]
```diff
@@ -3890,8 +3890,11 @@ def inner():
             results.append(True)
             return True
 
-        self.assertRaises(NotImplementedError, inner)
-        self.assertFalse(results)
+        with pytest.raises(NotImplementedError) as e:
+            inner()
+
+        assert "Error text" in str(e.value)
+        assert not results
```
Comment (hynek) [inline tests/test_ssl.py]: not is a bit too vague…[] == results is what you mean, right?

File: tests/test_ssl.py [hunk 5]
```diff
@@ -3874,8 +3874,8 @@ def inner():
             results.append(True)
             return True
 
-        self.assertTrue(inner())
-        self.assertEqual(results, [True])
+        assert inner()
```
Comment (hynek) [inline tests/test_ssl.py]: `is True` is nicer and more explicit

**Comment by hynek** [comment]
Thanks.  It’s stupid that the coverage check fails because the coverage decreased because we have less LoC…

**Pull request merged**